### ISSUE TITLE
Multiple resultset support

### ISSUE THREAD

**Issue by edefimov** [doctrine/dbal#3646]
### Feature Request

Some databases has native support of returning multiple result-sets within single query, for example Microsoft SQL Server. I suggest to add the feature into DBAL.

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

The feature may be implemented as a method of `Doctrine\DBAL\Driver\Statement`, or for keeping BC as a new interface implemented on statements, that support multiple result sets. It can look like this:
```php
$statement = $connection->executeQuery(...);
$result = [];
do {
    $result[] = $statement->fetchAll();
} while ($statement->hasNextRowSet());
```

**Comment by Ocramius** [comment]
Unless this is a cross-platform feature, this should not land in DBAL

**Comment by edefimov** [comment]
I think actually not only SQL server supports this, oracle can do it, postgresql also. The feature is already implemented in PDO and can be used in a way like:
```php
$statement = $connection->executeQuery(...);
$result = [];
do {
    $result[] = $statement->fetchAll();
} while ($statement instanceof \PDOStatement && $statement->nextRowset());
```
But in the case when database driver is not PDO, feature is not available.

**Comment by morozov** [comment]
@edefimov could you provide an example of a query which returns multiple result sets and a practical example why such queries should be used?

**Comment by edefimov** [comment]
As an example of such a query - a stored procedure. 
<details>
<summary>Sample SQL server code</summary>

```
CREATE TABLE integers (
    id INT PRIMARY KEY,
    value INT
);

CREATE TABLE doubles (
    id INT PRIMARY KEY,
    value DOUBLE PRECISION
);

INSERT INTO integers VALUES (1,1), (2,2), (3,3);
INSERT INTO doubles VALUES (1,1.0), (2,2.0), (3,3.0);

CREATE OR ALTER PROCEDURE multi_results
AS
BEGIN
    SET NOCOUNT, XACT_ABORT ON

    SELECT * FROM integers;
    SELECT * FROM doubles;
END
go
```
</details>

```php
$statement = $connection->executeQuery('EXEC multi_results');
```
A more practical usage: there is a reporting system which should be integrated with PHP application. The reporting system is written using stored procedures in database server. Some of them return multiple results and that behavior can not be changed in reasonable time. PHP application should call stored procedures and somehow process results.

**Comment by Ocramius** [comment]
I'd say that's not a valid use-case. If you go that deep into custom storage layer features, using the connector directly is a better idea. DBAL is already known to not work correctly with stored procedures at all times (in/out parameters, cursors, SP that do/do not produce resultsets conditionally).

At that point, the `AL` in `DBAL` is no longer useful, but rather a hindrance.

**Comment by edefimov** [comment]
May be simply add a method into SqlSrvStatement:
```php
public function nextRowset()
{
    $result = sqlsrv_next_result($this->stmt);
    if ($result === false) {
        throw SQLSrvException::fromSqlSrvErrors();
    }

    return $result === true;
}
```

**Comment by morozov** [comment]
It should be solved in the way agreed upon in #3534.
